packages/remirror__extension-highlight/src/utils/joined-highlights.ts

```diff
@@ -0,0 +1,51 @@
+import { HighlightAttrs } from '../types';
+
+/**
+ * Util function to partition an array of highlights into two
+ * sub-arrays based on the passed callback `filter`
+ */
+function partitionHighlights<HighlightAttrs>(
+  array: HighlightAttrs[],
+  filter: (val: HighlightAttrs) => boolean,
+): [HighlightAttrs[], HighlightAttrs[]] {
+  const matches = array.filter((e) => filter(e));
+  const notMatches = array.filter((e) => !filter(e));
+  return [matches, notMatches];
+}
+
+const joinDisjoinedHighlight = (highlights: HighlightAttrs[], highlight: HighlightAttrs) => {
+  let { id, from, to, text } = highlight;
+
+  if (from >= to) {
+    return highlights; // Sanity check.
+  }
+
+  // Find outer bound of all marks belong to the highlight
+  const [same, diff] = partitionHighlights(highlights, (h) => h.id === id);
+  from = Math.min(...same.map((h) => h.from), from);
+  to = Math.max(...same.map((h) => h.to), to);
+  let fullText = same.map((h) => h.text).join(' ');
+  // Respect existing keys and merge them into the new highlight.
+  const newHighlight: HighlightAttrs = {
+    ...highlight,
+    from,
+    to,
+    text: (fullText += text),
+  };
+  return [...diff, newHighlight];
+};
+
+/**
+ * @returns highlights joined by ID
+ */
+export const joinDisjoinedHighlights = (
+  disjoinedHighlights: HighlightAttrs[],
+): HighlightAttrs[] => {
+  const joinedHighlights: HighlightAttrs[] = [];
+
+  for (const disjoinedHighlight of disjoinedHighlights) {
+    joinDisjoinedHighlight(joinedHighlights, disjoinedHighlight);
+  }
+
+  return joinedHighlights;
+};
```

Comment:
This seem very convoluted, you could achieve the same by looking through node and interrogating each text node.

[see here](https://codesandbox.io/s/intelligent-stonebraker-00vgkt?file=/src/HighlightExtension.ts:1738-2785)

Reply:
Thank you for the insight. It's true that it looks convoluted. The idea behind getting the disjoined marks is to create decorations based on them (e.g blending the colors of the overlapping marks). overlapping marks, in this case, are represented as a subArray in the return value of `getDisjoinedEntityReferences` helper. I guess we could somehow "disjoin" them using the overlapping ranges after we get the _joined marks_ (using `extractHighlightMarks` from the sandbox). Do you think that's an optimal way to do it?